### PR TITLE
Documentation fixes/improvements

### LINKED ISSUE #17817: Fix iOS documentation build
##### System information (version)
- OpenCV => master
- Operating System / Platform => iOS
- Compiler => XCode

##### Detailed description

Build fails with:
Fatal error: Unsupported CXCursorKind: UnionDecl: file /Users/johnf/project/jazzy/SourceKitten/Source/SourceKittenFramework/ObjCDeclarationKind.swift

##### Steps to reproduce

Install jazzy (`gem install jazzy`)
Run iOS build (`./opencv/platforms/ios/build_framework.py iosbuild`)

##### Issue submission checklist

 - [x] I report the issue, it's not a question
 - [x] I updated to latest OpenCV version and the issue is still there

I have a PR ready for this

### PR DESCRIPTION
**Pull request by komakai**
This PR
 - fixes #17817
 - improves the documentation to show the relationship between classes and modules (partially addresses comment from @vpisarev on PR #17165 )
 - adds favicon to documentation bundle

<cut/>

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under OpenCV (BSD) License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work

### CODE REVIEW

**Comment by komakai** [comment]
Preview of update docs:
http://xtravision.stars.ne.jp/opencv-objc-doc-test/docs/index.html

**Comment by vpisarev** [comment]
@komakai, thank you for another great contribution! I checked the docs and indeed they look much better!

File: modules/core/misc/objc/common/CVObjcUtil.h [hunk 1]
```diff
@@ -6,11 +6,13 @@
 
 #pragma once
 
+#ifdef HASH_UTILS
```
Comment (vpisarev) [inline modules/core/misc/objc/common/CVObjcUtil.h:9]: I wonder, if those DOUBLE_TO_BITS, FLOAT_TO_BITS is necessary to have in the form of macros in header files? I guess, those conversions are not that performance-critical? Can we just declare those functions in the header file and implement, together with unions (btw, there are already such unions, Cv32suf and Cv64suf) somewhere in .cpp or .mm file? In this case, we don't have to put HASH_UTILS define in many headers

Comment (komakai) [reply modules/core/misc/objc/common/CVObjcUtil.h]: @vpisarev thanks for pointing out about Cv32suf and Cv64suf - I have modified the PR to use them and removed the redundant V64 and V32 unions. I also moved the DOUBLE_TO_BITS and FLOAT_TO_BITS macros out to just the files that need them.

**Comment by vpisarev** [comment]
@komakai, thank you!

I've tried it out, the new docs build well and looks much better!

The only thing that, well, does not work is search. I upgraded to the latest jazzy 0.13.5, and the search window is there, in the top, but if I type something there and press enter, I simply get a huge json file. Do you happen to know what can be the reason?

**Comment by komakai** [comment]
> the search window is there, on the top, but if I type something there and press enter, I simply get a huge json file. Do you happen to know what can be the reason?

I think the problem is your browser has disabled javascript for files loaded directly from the file system. If you open a terminal and go to the `docs` folder and run `python -m SimpleHTTPServer` then browse to `http://localhost:8000/` everything should work

**Comment by vpisarev** [comment]
@komakai, thank you, with such a trick it works well! :+1:

**Comment by komakai** [comment]
@alalek Any chance you could add a link to the Objective-C/Swift document from the OpenCV documentation page?
ObjC-SwiftDoc


Thanks 😃

**Comment by alalek** [comment]
@komakai It makes sense.
Do you know if it is possible to build Obj-C/Swift documentation on Linux environment?

**Comment by komakai** [comment]
> Do you know if it is possible to build Obj-C/Swift documentation on Linux environment?

@alalek let me check and see if there is a way to do it - looking at https://gist.github.com/jpsim/eed98cc50f9c979ebfbce7616bb4e161 it seems like it might be possible

**Comment by komakai** [comment]
> Do you know if it is possible to build Obj-C/Swift documentation on Linux environment?

@alalek Unfortunately not. Fails here:
https://github.com/jpsim/SourceKitten/blob/662dad97eb4c94158b8f24160a7d539a039126d4/Source/sourcekitten/Doc.swift#L30

**Pull request merged**